flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java

```diff
@@ -523,6 +545,103 @@ public JobID getOwningJob(AllocationID allocationId) {
         }
     }
 
+    private void tryPersistAllocationSnapshot(
+            int requestedIndex,
+            JobID jobId,
+            AllocationID allocationId,
+            ResourceProfile resourceProfile) {
+        if (!allocationsDirectory.exists() && !allocationsDirectory.mkdirs()) {
+            LOG.debug(
+                    "Allocations directory {} doesn't exist and cannot be created.",
+                    allocationsDirectory.toPath());
+        }
+
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile =
+                new File(allocationsDirectory.getAbsolutePath(), requestedIndex + ".bin");
+        try (ObjectOutputStream oos =
+                new ObjectOutputStream(new FileOutputStream(slotAllocationSnapshotFile))) {
+            oos.writeObject(
+                    new LocalSlotAllocationSnapshot(
+                            requestedIndex, jobId, allocationId, resourceProfile));
+            LOG.debug(
+                    "Successfully written allocation state metadata file {} for job {} and allocation {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    jobId,
+                    allocationId);
+        } catch (IOException e) {
+            // TODO is this ok or should this hard fail?
+            LOG.debug(
+                    "Cannot write the local allocations state. File {} for job {} and allocation {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    jobId,
+                    allocationId);
+        }
+    }
+
+    private void tryCleanupSlotAllocationSnapshot(int requestedIndex) {
+        if (!allocationsDirectory.exists()) {
+            LOG.debug(
+                    "There is no local allocations snapshot directory to cleanup {}.",
+                    allocationsDirectory.toPath());
+        }
+
+        // Let's try to write the slot allocations on file
+        final File slotAllocationSnapshotFile =
+                new File(allocationsDirectory.getAbsolutePath(), requestedIndex + ".bin");
+        try {
+            FileUtils.deleteFileOrDirectory(slotAllocationSnapshotFile);
+            LOG.debug(
+                    "Successfully deleted allocation state metadata file {}.",
+                    slotAllocationSnapshotFile.toPath());
+        } catch (IOException e) {
+            LOG.debug(
+                    "Cannot delete the local allocations state file {}.",
+                    slotAllocationSnapshotFile.toPath(),
+                    e);
+        }
+    }
+
+    private void tryReloadLocalAllocationSnapshots(Time slotTimeout) {
+        if (!allocationsDirectory.exists()) {
+            LOG.debug(
+                    "There is no local allocations snapshot directory to load from {}.",
+                    allocationsDirectory.toPath());
+        }
```

Comment:
I would establish the invariant that `allocationsDirectory` exists when initializing this class.